crates/subspace-farmer/src/dsn.rs

```diff
@@ -154,47 +153,40 @@ impl DSNSync for subspace_networking::Node {
 
         // obtain closest peers to the middle of the range
         let key = Code::Identity.digest(&middle);
-        let peers = self
-            .get_closest_peers(key)
-            .await
-            .map_err(GetPiecesByRangeError::CannotGetClosestPeers)?;
-
-        // select appropriate peer for the piece-by-range protocol
-        if peers.is_empty() {
-            return Err(GetPiecesByRangeError::NoClosestPiecesFound);
-        }
-
-        let mut peer_id: Option<PeerId> = None;
-        for peer_id_candidate in peers {
-            let result = self
-                .send_generic_request(peer_id_candidate, PeerInfoRequest)
-                .await;
-
-            trace!(
-                peer_id = %peer_id_candidate,
-                "Peer info request returned {:?}",
-                result
-            );
+        let peer_id = {
+            let node = self.clone();
+
+            backoff::future::retry(
+                backoff::ExponentialBackoffBuilder::new()
+                    .with_max_elapsed_time(Some(std::time::Duration::from_secs(1)))
+                    .build(),
+                move || {
+                    let node = node.clone();
+                    async move {
+                        let peers = node
+                            .get_closest_peers(key)
+                            .await
+                            .map_err(GetPiecesByRangeError::CannotGetClosestPeers)?;
+                        tracing::debug!(?peers, "Received peers");
+
+                        // select first peer for the piece-by-range protocol
+                        for id in peers.into_iter().take(5) {
+                            if node
+                                .send_generic_request(id, PeerInfoRequest)
+                                .await
+                                .map(|response| response.peer_info.status == PeerSyncStatus::Ready)
+                                .unwrap_or_default()
+                            {
+                                return Ok(id);
+                            }
+                        }
 
-            match result {
-                Ok(response) => {
-                    if response.peer_info.status == PeerSyncStatus::Ready {
-                        peer_id = Some(peer_id_candidate);
-                    } else {
-                        trace!(
-                            peer_id = %peer_id_candidate,
-                            status = ?response.peer_info.status,
-                            "Peer was not selected for synchronization."
-                        );
+                        Err(GetPiecesByRangeError::NoSupportedFarmer.into())
```

Comment:
Consider use transient or permanent errors explicitly here. Permanent means retries will not help and need to be stopped, transient means retries might help and more attempts should be done.

The default conversion is to transient errors, but making it explicit is probably beneficial for readability.